src/Data/Primitive/Contiguous/Class.hs

```diff
@@ -0,0 +1,992 @@
+  -- they are not really meant for direct use.
+  --
+  -- @since 0.6.0
+  thaw_ :: (PrimMonad m, Element arr b)
+    => arr b
+    -> Int -- ^ offset into the array
+    -> Int -- ^ length of the slice
+    -> m (Mutable arr (PrimState m) b)
+
+  ------ Copy Operations ------
+  -- | Copy a slice of an array into a mutable array.
+  copy :: (PrimMonad m, Element arr b)
+    => Mutable arr (PrimState m) b -- ^ destination array
+    -> Int -- ^ offset into destination array
+    -> Sliced arr b -- ^ source slice
+    -> m ()
+  default copy ::
+      ( Sliced arr ~ Slice arr, ContiguousU arr
+      , PrimMonad m, Element arr b)
+    => Mutable arr (PrimState m) b -> Int -> Sliced arr b -> m ()
+  {-# INLINE copy #-}
+  copy dst dstOff Slice{offset,length,base} = copy_ dst dstOff (lift base) offset length
+  -- | Copy a slice of an array into a mutable array without using the 'Slice' type.
+  -- These methods are required to implement 'Contiguous (Slice arr)' for any `Contiguous arr`;
+  -- they are not really meant for direct use.
+  --
+  -- @since 0.6.0
+  copy_ :: (PrimMonad m, Element arr b)
+    => Mutable arr (PrimState m) b -- ^ destination array
+    -> Int -- ^ offset into destination array
+    -> arr b -- ^ source array
+    -> Int -- ^ offset into source array
+    -> Int -- ^ number of elements to copy
+    -> m ()
+  -- | Copy a slice of a mutable array into another mutable array.
+  --   In the case that the destination and source arrays are the
+  --   same, the regions may overlap.
+  copyMut :: (PrimMonad m, Element arr b)
+    => Mutable arr (PrimState m) b -- ^ destination array
+    -> Int -- ^ offset into destination array
+    -> MutableSliced arr (PrimState m) b -- ^ source slice
+    -> m ()
+  default copyMut ::
+       ( MutableSliced arr ~ MutableSlice arr, ContiguousU arr
+       , PrimMonad m, Element arr b)
+    => Mutable arr (PrimState m) b -> Int -> MutableSliced arr (PrimState m) b -> m ()
+  {-# INLINE copyMut #-}
+  copyMut dst dstOff MutableSlice{offsetMut,lengthMut,baseMut}
+    = copyMut_ dst dstOff (liftMut baseMut) offsetMut lengthMut
+  -- | Copy a slice of a mutable array into another mutable array without using the 'Slice' type.
+  -- These methods are required to implement 'Contiguous (Slice arr)' for any `Contiguous arr`;
+  -- they are not really meant for direct use.
+  --
+  -- @since 0.6.0
+  copyMut_ :: (PrimMonad m, Element arr b)
+    => Mutable arr (PrimState m) b -- ^ destination array
+    -> Int -- ^ offset into destination array
+    -> Mutable arr (PrimState m) b -- ^ source array
+    -> Int -- ^ offset into source array
+    -> Int -- ^ number of elements to copy
+    -> m ()
+  -- | Copy a slice of an array and then insert an element into that array.
+  --
+  -- The default implementation performs a memset which would be unnecessary
+  -- except that the garbage collector might trace the uninitialized array.
+  insertSlicing :: (Element arr b)
```

Comment:
The name `insertSlicing` was always unfortunate. It used to be important to call it this because it drew attention to the function's unusual behavior of combining two unrelated operations, all for the glorious cause of avoiding allocation in an intermediate array. However, this function doesn't actually slice anymore, which is wonderful! So, it can rightfully be renamed to `insert`.